tests/robustness/watch.go

```diff
@@ -131,142 +132,139 @@ func watchResponsesMaxRevision(responses []traffic.WatchResponse) int64 {
 	var maxRevision int64
 	for _, response := range responses {
 		for _, event := range response.Events {
-			if event.Kv.ModRevision > maxRevision {
-				maxRevision = event.Kv.ModRevision
+			if event.Revision > maxRevision {
+				maxRevision = event.Revision
 			}
 		}
 	}
 	return maxRevision
 }
 
-func validateWatchResponses(t *testing.T, clus *e2e.EtcdProcessCluster, responses [][]traffic.WatchResponse, expectProgressNotify bool) {
-	for i, member := range clus.Procs {
-		validateMemberWatchResponses(t, member.Config().Name, responses[i], expectProgressNotify)
+func validateWatchCorrectness(t *testing.T, reports []traffic.ClientReport) {
+	// Validate etcd watch properties defined in https://etcd.io/docs/v3.6/learning/api_guarantees/#watch-apis
+	for _, r := range reports {
+		validateOrdered(t, r)
+		validateUnique(t, r)
+		validateAtomic(t, r)
+		validateBookmarkable(t, r)
+		// TODO: Validate presumable
 	}
+	validateEventsMatch(t, reports)
+	// Expects that longest history encompasses all events.
+	// TODO: Use combined events from all histories instead of the longest history.
+	// TODO: Validate that each watch report is reliable, not only the longest one.
+	validateReliable(t, longestEventHistory(reports))
 }
 
-func validateMemberWatchResponses(t *testing.T, memberId string, responses []traffic.WatchResponse, expectProgressNotify bool) {
-	// Validate watch is correctly configured to ensure proper testing
-	validateGotAtLeastOneProgressNotify(t, memberId, responses, expectProgressNotify)
-
-	// Validate etcd watch properties defined in https://etcd.io/docs/v3.6/learning/api/#watch-streams
-	validateOrderedAndReliable(t, memberId, responses)
-	validateUnique(t, memberId, responses)
-	validateAtomic(t, memberId, responses)
-	// Validate kubernetes usage of watch
-	validateRenewable(t, memberId, responses)
-}
-
-func validateGotAtLeastOneProgressNotify(t *testing.T, memberId string, responses []traffic.WatchResponse, expectProgressNotify bool) {
+func validateGotAtLeastOneProgressNotify(t *testing.T, reports []traffic.ClientReport, expectProgressNotify bool) {
```

Comment:
Is it expected only one out of x watch clients get one progress notify. 

I think all x watch clients should get one, respectively because otherwise, that does not sound right from k8s prospective.

Reply:
That's not right, manual progress notify is sent back to client requesting it.

Reply:
https://github.com/etcd-io/etcd/blob/efcbd95f71a131953ceb1ea474256399af8182ea/server/etcdserver/api/v3rpc/watch.go#L322 cc @chaochn47

Reply:
I see. the progress notify contains manual requested and periodic. 

Is my statement true for the periodic one?

Reply:
Only if all watchers watch a event stream that didn't have any event for required amount of time.